pkg/sql/schemachanger/sctest/cumulative.go

```diff
@@ -37,13 +40,428 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/datadriven"
 	"github.com/cockroachdb/errors"
 	"github.com/lib/pq"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/sync/errgroup"
 )
 
+type stageExecType int
+
+const (
+	_                 stageExecType = iota
+	stageExecuteQuery stageExecType = 1
+	stageExecuteStmt  stageExecType = 2
+)
+
+// stageExecStmt a statement that will be executed during a given stage,
+// including any expected errors from this statement or any schema change
+// running concurrently.
+type stageExecStmt struct {
+	execType               stageExecType
+	stmts                  []string
+	expectedOutput         string
+	observedOutput         string
+	schemaChangeErrorRegex *regexp.Regexp
+}
+
+// HasSchemaChangeError will the execution of this statement lead to a schema
+// change error.
+func (e *stageExecStmt) HasSchemaChangeError() bool {
+	return e.schemaChangeErrorRegex != nil
+}
+
+// stageKeyOrdinalLatest targets the latest ordinal in a stage.
+const stageKeyOrdinalLatest = math.MaxUint16
+
+// stageKey represents a phase and stage range to target (in an
+// inclusive manner).
+type stageKey struct {
+	minOrdinal int
+	maxOrdinal int
+	phase      scop.Phase
+}
+
+// makeStageKey constructs a stage key targeting a single ordinal.
+func makeStageKey(phase scop.Phase, ordinal int) stageKey {
+	return stageKey{
+		phase:      phase,
+		minOrdinal: ordinal,
+		maxOrdinal: ordinal,
+	}
+}
+
+// AsInt converts the stage index into a unique numeric integer.
+func (s *stageKey) AsInt() int {
+	// Assuming we never have plans with more than 1000 stages per-phase.
+	return (int(s.phase) * 1000) + s.minOrdinal
+}
+
+// String implements fmt.Stringer
+func (s *stageKey) String() string {
+	if s.minOrdinal == s.maxOrdinal {
+		return fmt.Sprintf("(phase = %s stageOrdinal=%d)",
+			s.phase, s.minOrdinal)
+	}
+	return fmt.Sprintf("(phase = %s stageMinOrdinal=%d stageMaxOrdinal=%d),",
+		s.phase, s.minOrdinal, s.maxOrdinal)
+}
+
+// IsEmpty detects if a stage key is empty.
+func (s *stageKey) IsEmpty() bool {
+	return s.phase == 0
+}
+
```

Comment:
nit: let's add a comment here that says something along the line of "a stageKeyEntry captures the stage range in which statement(s) is to be executed"